[CODE HEALTH] GCC reports warnings in maintainer Release builds

CI tests building in maintainer mode with GCC without optimizations. When built in `Release` new warnings are reported. 

Running the following test in the development container results in two new warnings. 
```
export BUILD_TYPE=Release
./ci/do_ci.sh cmake.maintainer.async.test
```

-----
**otlp_file_client.cc**
```
/workspaces/opentelemetry-cpp/exporters/otlp/src/otlp_file_client.cc: In member function 'std::shared_ptr<_IO_FILE> opentelemetry::v1::exporter::otlp::OtlpFileSystemBackend::OpenLogFile(bool)':
/workspaces/opentelemetry-cpp/exporters/otlp/src/otlp_file_client.cc:1259:19: error: ignoring return value of 'char* strerror_r(int, char*, size_t)' declared with attribute 'warn_unused_result' [-Werror=unused-result]
 1259 |         strerror_r(error_code, error_message, sizeof(error_message) - 1);
```

 -----
**circular_buffer.cc**
 ```
 In file included from /usr/include/c++/13/vector:66,
                 from /workspaces/opentelemetry-cpp/sdk/src/metrics/data/circular_buffer.cc:8:
In constructor 'std::_Vector_base<_Tp, _Alloc>::_Vector_impl_data::_Vector_impl_data(std::_Vector_base<_Tp, _Alloc>::_Vector_impl_data&&) [with _Tp = unsigned int; _Alloc = std::allocator<unsigned int>]',
    inlined from 'std::_Vector_base<_Tp, _Alloc>::_Vector_impl::_Vector_impl(std::_Vector_base<_Tp, _Alloc>::_Vector_impl&&) [with _Tp = unsigned int; _Alloc = std::allocator<unsigned int>]' at /usr/include/c++/13/bits/stl_vector.h:155:68,
    inlined from 'std::_Vector_base<_Tp, _Alloc>::_Vector_base(std::_Vector_base<_Tp, _Alloc>&&) [with _Tp = unsigned int; _Alloc = std::allocator<unsigned int>]' at /usr/include/c++/13/bits/stl_vector.h:338:7,
    inlined from 'std::vector<_Tp, _Alloc>::vector(std::vector<_Tp, _Alloc>&&) [with _Tp = unsigned int; _Alloc = std::allocator<unsigned int>]' at /usr/include/c++/13/bits/stl_vector.h:620:7,
    inlined from 'static typename absl::otel_v1::variant_alternative<I, T>::type& absl::otel_v1::variant_internal::VariantCoreAccess::Replace(Self*, Args&& ...) [with long unsigned int NewIndex = 2; Self = absl::otel_v1::variant<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >; Args = {std::vector<unsigned int, std::allocator<unsigned int> >}]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/../types/internal/variant.h:727:25,
    inlined from 'absl::otel_v1::variant_alternative_t<I, absl::otel_v1::variant<T0, Ts ...> >& absl::otel_v1::variant<T0, Ts ...>::emplace(Args&& ...) [with long unsigned int I = 2; Args = {std::vector<unsigned int, std::allocator<unsigned int> >}; typename std::enable_if<std::is_constructible<typename absl::otel_v1::variant_alternative<I, absl::otel_v1::variant<T0, Ts ...> >::type, Args ...>::value>::type* <anonymous> = 0; T0 = std::vector<unsigned char>; Tn = {std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> >}]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/variant.h:665:59,
    inlined from 'void absl::otel_v1::variant_internal::VariantCoreAccess::MoveAssignVisitor<VType>::operator()(absl::otel_v1::variant_internal::SizeT<NewIndex>) const [with long unsigned int NewIndex = 2; VType = absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/../types/internal/variant.h:628:50,
    inlined from 'static decltype (declval<F>()((declval<Args>)()...)) absl::otel_v1::base_internal::Callable::Invoke(F&&, Args&& ...) [with F = absl::otel_v1::variant_internal::VariantCoreAccess::MoveAssignVisitor<absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > > >; Args = {std::integral_constant<long unsigned int, 2>}]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/../utility/../base/internal/invoke.h:153:30,
    inlined from 'absl::otel_v1::base_internal::InvokeT<F, Args ...> absl::otel_v1::base_internal::Invoke(F&&, Args&& ...) [with F = absl::otel_v1::variant_internal::VariantCoreAccess::MoveAssignVisitor<absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > > >; Args = {std::integral_constant<long unsigned int, 2>}]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/../utility/../base/internal/invoke.h:180:43,
    inlined from 'static absl::otel_v1::variant_internal::VisitIndicesResultT<Op, long unsigned int> absl::otel_v1::variant_internal::ReachableSwitchCase<Op, I>::Run(Op&&) [with Op = absl::otel_v1::variant_internal::VariantCoreAccess::MoveAssignVisitor<absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > > >; long unsigned int I = 2]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/../types/internal/variant.h:296:69,
    inlined from 'static absl::otel_v1::variant_internal::VisitIndicesResultT<Op, long unsigned int> absl::otel_v1::variant_internal::VisitIndicesSwitch<EndIndex>::Run(Op&&, std::size_t) [with Op = absl::otel_v1::variant_internal::VariantCoreAccess::MoveAssignVisitor<absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > > >; long unsigned int EndIndex = 4]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/../types/internal/variant.h:365:46,
    inlined from 'absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<T>& absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<T>::operator=(absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<T>&&) [with T = {std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> >}]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/../types/internal/variant.h:1429:38,
    inlined from 'absl::otel_v1::variant_internal::VariantCopyAssignBaseNontrivial<T>& absl::otel_v1::variant_internal::VariantCopyAssignBaseNontrivial<T>::operator=(absl::otel_v1::variant_internal::VariantCopyAssignBaseNontrivial<T>&&) [with T = {std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> >}]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/../types/internal/variant.h:1453:36,
    inlined from 'absl::otel_v1::variant<T0, Ts ...>& absl::otel_v1::variant<T0, Ts ...>::operator=(absl::otel_v1::variant<T0, Ts ...>&&) [with T0 = std::vector<unsigned char>; Tn = {std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> >}]' at /workspaces/opentelemetry-cpp/api/include/opentelemetry/nostd/internal/absl/types/variant.h:577:12,
    inlined from 'std::_Require<std::__not_<std::__is_tuple_like<_Tp> >, std::is_move_constructible<_Tp>, std::is_move_assignable<_Tp> > std::swap(_Tp&, _Tp&) [with _Tp = absl::otel_v1::variant<vector<unsigned char, allocator<unsigned char> >, vector<short unsigned int, allocator<short unsigned int> >, vector<unsigned int, allocator<unsigned int> >, vector<long unsigned int, allocator<long unsigned int> > >]' at /usr/include/c++/13/bits/move.h:199:11,
    inlined from 'void opentelemetry::v1::sdk::metrics::AdaptingIntegerArray::EnlargeToFit(uint64_t)' at /workspaces/opentelemetry-cpp/sdk/src/metrics/data/circular_buffer.cc:126:12:
/usr/include/c++/13/bits/stl_vector.h:107:11: error: '*(const std::_Vector_base<long unsigned int, std::allocator<long unsigned int> >::_Vector_impl_data*)((char*)&__tmp + offsetof(absl::otel_v1::variant<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >,absl::otel_v1::variant<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantCopyAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantCopyBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantMoveBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantStateBaseDestructorNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantStateBase<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::state_)).std::_Vector_base<long unsigned int, std::allocator<long unsigned int> >::_Vector_impl_data::_M_end_of_storage' may be used uninitialized [-Werror=maybe-uninitialized]
  107 |           _M_end_of_storage(__x._M_end_of_storage)
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/c++/13/bits/stl_pair.h:61,
                 from /usr/include/c++/13/utility:69,
                 from /workspaces/opentelemetry-cpp/sdk/src/metrics/data/circular_buffer.cc:7:
/usr/include/c++/13/bits/move.h: In member function 'void opentelemetry::v1::sdk::metrics::AdaptingIntegerArray::EnlargeToFit(uint64_t)':
/usr/include/c++/13/bits/move.h:197:11: note: '*(const std::_Vector_base<long unsigned int, std::allocator<long unsigned int> >::_Vector_impl_data*)((char*)&__tmp + offsetof(absl::otel_v1::variant<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >,absl::otel_v1::variant<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantCopyAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantMoveAssignBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantCopyBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantMoveBaseNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantStateBaseDestructorNontrivial<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::<unnamed>.absl::otel_v1::variant_internal::VariantStateBase<std::vector<unsigned char, std::allocator<unsigned char> >, std::vector<short unsigned int, std::allocator<short unsigned int> >, std::vector<unsigned int, std::allocator<unsigned int> >, std::vector<long unsigned int, std::allocator<long unsigned int> > >::state_)).std::_Vector_base<long unsigned int, std::allocator<long unsigned int> >::_Vector_impl_data::_M_end_of_storage' was declared here
  197 |       _Tp __tmp = _GLIBCXX_MOVE(__a);
 ```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -104,14 +104,14 @@ jobs:
     - name: install dependencies
       run: |
         sudo -E ./ci/install_thirdparty.sh --install-dir /usr/local --tags-file third_party_release
-    - name: run cmake gcc (maintainer mode, sync, Release)
-      env:
-        BUILD_TYPE: Release
-      run: |
-        ./ci/do_ci.sh cmake.maintainer.sync.test
     - name: run cmake gcc (maintainer mode, sync, Debug)
       env:
         BUILD_TYPE: Debug
+      run: |
+        ./ci/do_ci.sh cmake.maintainer.sync.test
+    - name: run cmake gcc (maintainer mode, sync, Release)
+      env:
+        BUILD_TYPE: Release
       run: |
         ./ci/do_ci.sh cmake.maintainer.sync.test
     - name: generate test cert
@@ -142,14 +142,14 @@ jobs:
     - name: install dependencies
       run: |
         sudo -E ./ci/install_thirdparty.sh --install-dir /usr/local --tags-file third_party_release
-    - name: run cmake gcc (maintainer mode, async, Release)
-      env:
-        BUILD_TYPE: Release
-      run: |
-        ./ci/do_ci.sh cmake.maintainer.async.test
     - name: run cmake gcc (maintainer mode, async, Debug)
       env:
         BUILD_TYPE: Debug
+      run: |
+        ./ci/do_ci.sh cmake.maintainer.async.test
+    - name: run cmake gcc (maintainer mode, async, Release)
+      env:
+        BUILD_TYPE: Release
       run: |
         ./ci/do_ci.sh cmake.maintainer.async.test
     - name: generate test cert

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -104,7 +104,14 @@ jobs:
     - name: install dependencies
       run: |
         sudo -E ./ci/install_thirdparty.sh --install-dir /usr/local --tags-file third_party_release
-    - name: run cmake gcc (maintainer mode, sync)
+    - name: run cmake gcc (maintainer mode, sync, Release)
+      env:
+        BUILD_TYPE: Release
+      run: |
+        ./ci/do_ci.sh cmake.maintainer.sync.test
+    - name: run cmake gcc (maintainer mode, sync, Debug)
+      env:
+        BUILD_TYPE: Debug
       run: |
         ./ci/do_ci.sh cmake.maintainer.sync.test
     - name: generate test cert
@@ -135,7 +142,14 @@ jobs:
     - name: install dependencies
       run: |
         sudo -E ./ci/install_thirdparty.sh --install-dir /usr/local --tags-file third_party_release
-    - name: run cmake gcc (maintainer mode, async)
+    - name: run cmake gcc (maintainer mode, async, Release)
+      env:
+        BUILD_TYPE: Release
+      run: |
+        ./ci/do_ci.sh cmake.maintainer.async.test
+    - name: run cmake gcc (maintainer mode, async, Debug)
+      env:
+        BUILD_TYPE: Debug
       run: |
         ./ci/do_ci.sh cmake.maintainer.async.test
     - name: generate test cert

--- a/exporters/otlp/src/otlp_file_client.cc
+++ b/exporters/otlp/src/otlp_file_client.cc
@@ -1256,7 +1256,8 @@ private:
         strerror_s(error_message, sizeof(error_message) - 1, error_code);
 #else
         char error_message[256] = {0};
-        strerror_r(error_code, error_message, sizeof(error_message) - 1);
+        OPENTELEMETRY_MAYBE_UNUSED auto strerror_r_result =
+            strerror_r(error_code, error_message, sizeof(error_message) - 1);
 #endif
         OTEL_INTERNAL_LOG_ERROR("[OTLP FILE Client] Create directory \""
                                 << directory_name << "\" failed.errno: " << error_code

--- a/sdk/src/metrics/data/circular_buffer.cc
+++ b/sdk/src/metrics/data/circular_buffer.cc
@@ -123,8 +123,8 @@ void AdaptingIntegerArray::EnlargeToFit(uint64_t value)
   {
     backing = std::vector<uint64_t>(backing_size, 0);
   }
-  std::swap(backing_, backing);
-  nostd::visit(AdaptingIntegerArrayCopy{}, backing, backing_);
+  nostd::visit(AdaptingIntegerArrayCopy{}, backing_, backing);
+  backing_ = std::move(backing);
 }
 
 void AdaptingCircularBufferCounter::Clear()

